File: edge/mesh.py
"""
Mesh networking module for EchoSense edge devices.

Implements a simple AODV-like routing protocol for peer-to-peer communication
between edge devices in a wireless mesh network.
"""

import time
import random
import struct
import threading
import socket
import json
from typing import Dict, List, Optional, Tuple, Set
from dataclasses import dataclass, field
from enum import Enum
import logging

logger = logging.getLogger("mesh")


class MessageType(Enum):
    """Message types for mesh protocol."""
    DISCOVERY = 0x01
    ROUTE_REQUEST = 0x02
    ROUTE_REPLY = 0x03
    DATA = 0x04
    ACK = 0x05


@dataclass
class Neighbor:
    """Represents a neighboring device in the mesh network."""
    device_id: str
    ip: str
    port: int
    last_seen: float = field(default_factory=time.time)
    signal_strength: float = 0.0  # RSSI in dBm
    hop_count: int = 1


@dataclass
class RouteEntry:
    """Represents a route to a destination device."""
    destination: str
    next_hop: str
    hop_count: int
    sequence_number: int = 0
    last_used: float = field(default_factory=time.time)
    expires: float = field(default_factory=lambda: time.time() + 300)  # 5 min TTL


@dataclass
class MeshMessage:
    """A message in the mesh network."""
    msg_type: MessageType
    source: str
    destination: str
    sequence_number: int
    hop_count: int = 0
    max_hops: int = 10
    payload: bytes = b""
    previous_hop: Optional[str] = None

    def to_bytes(self) -> bytes:
        """Serialize message to bytes."""
        header = struct.pack(
            "!B4s4sIHHH",
            self.msg_type.value,
            self.source.encode("utf-8")[:4].ljust(4, b"\x00"),
            self.destination.encode("utf-8")[:4].ljust(4, b"\x00"),
            self.sequence_number,
            self.hop_count,
            self.max_hops,
            len(self.payload)
        )
        return header + self.payload

    @classmethod
    def from_bytes(cls, data: bytes) -> "MeshMessage":
        """Deserialize message from bytes."""
        if len(data) < 17:
            raise ValueError("Message too short")
        
        msg_type, src_bytes, dst_bytes, seq_num, hop_count, max_hops, payload_len = struct.unpack(
            "!B4s4sIHHH", data[:17]
        )
        
        source = src_bytes.decode("utf-8", errors="ignore").strip("\x00")
        destination = dst_bytes.decode("utf-8", errors="ignore").strip("\x00")
        payload = data[17:17+payload_len]
        
        return cls(
            msg_type=MessageType(msg_type),
            source=source,
            destination=destination,
            sequence_number=seq_num,
            hop_count=hop_count,
            max_hops=max_hops,
            payload=payload
        )


class MeshNetwork:
    """
    Mesh networking manager for edge devices.
    
    Implements neighbor discovery, route discovery, and packet forwarding.
    """
    
    def __init__(self, device_id: str, listen_port: int = 9000):
        self.device_id = device_id
        self.listen_port = listen_port
        self.neighbors: Dict[str, Neighbor] = {}
        self.routes: Dict[str, RouteEntry] = {}
        self.sequence_number = 0
        self.pending_routes: Dict[str, float] = {}  # destination -> request time
        self.message_queue: List[Tuple[str, bytes]] = []  # (destination, payload)
        
        self.running = False
        self.listen_thread: Optional[threading.Thread] = None
        self.discovery_thread: Optional[threading.Thread] = None
        
        # Socket for mesh communication
        self.socket: Optional[socket.socket] = None
        
    def start(self):
        """Start the mesh network manager."""
        if self.running:
            return
        
        self.running = True
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.socket.bind(("0.0.0.0", self.listen_port))
        self.socket.settimeout(1.0)
        
        # Start listener thread
        self.listen_thread = threading.Thread(target=self._listen_loop, daemon=True)
        self.listen_thread.start()
        
        # Start discovery thread
        self.discovery_thread = threading.Thread(target=self._discovery_loop, daemon=True)
        self.discovery_thread.start()
        
        logger.info(f"[MESH] Started on port {self.listen_port}")
        
    def stop(self):
        """Stop the mesh network manager."""
        self.running = False
        if self.socket:
            self.socket.close()
        logger.info("[MESH] Stopped")
        
    def add_neighbor(self, device_id: str, ip: str, port: int, signal_strength: float = 0.0):
        """Add or update a neighbor."""
        neighbor = Neighbor(
            device_id=device_id,
            ip=ip,
            port=port,
            last_seen=time.time(),
            signal_strength=signal_strength
        )
        self.neighbors[device_id] = neighbor
        logger.debug(f"[MESH] Added/updated neighbor: {device_id} at {ip}:{port}")
        
    def _listen_loop(self):
        """Listen for incoming mesh messages."""
        while self.running:
            try:
                data, addr = self.socket.recvfrom(1024)
                self._handle_message(data, addr)
            except socket.timeout:
                continue
            except Exception as e:
                logger.error(f"[MESH] Listen error: {e}")
                
    def _discovery_loop(self):
        """Periodically send discovery beacons."""
        while self.running:
            self._send_discovery()
            self._cleanup_stale_neighbors()
            self._cleanup_expired_routes()
            time.sleep(10)  # Discovery interval
            
    def _send_discovery(self):
        """Send a discovery beacon to broadcast address."""
        msg = MeshMessage(
            msg_type=MessageType.DISCOVERY,
            source=self.device_id,
            destination="BROADCAST",
            sequence_number=self._get_sequence_number(),
            payload=b""
        )
        self._broadcast(msg.to_bytes())
        logger.debug("[MESH] Sent discovery beacon")
        
    def _broadcast(self, data: bytes):
        """Broadcast message to all known neighbors."""
        for neighbor in self.neighbors.values():
            try:
                self.socket.sendto(data, (neighbor.ip, neighbor.port))
            except Exception as e:
                logger.warning(f"[MESH] Failed to send to {neighbor.device_id}: {e}")
                
    def _handle_message(self, data: bytes, addr: Tuple[str, int]):
        """Handle incoming mesh message."""
        try:
            msg = MeshMessage.from_bytes(data)
        except Exception as e:
            logger.error(f"[MESH] Failed to parse message: {e}")
            return
        
        # Update neighbor if this is from a direct neighbor
        if msg.source != self.device_id and msg.hop_count == 0:
            self.add_neighbor(msg.source, addr[0], addr[1])
        
        # Handle based on message type
        if msg.msg_type == MessageType.DISCOVERY:
            self._handle_discovery(msg, addr)
        elif msg.msg_type == MessageType.ROUTE_REQUEST:
            self._handle_route_request(msg, addr)
        elif msg.msg_type == MessageType.ROUTE_REPLY:
            self._handle_route_reply(msg, addr)
        elif msg.msg_type == MessageType.DATA:
            self._handle_data(msg, addr)
        elif msg.msg_type == MessageType.ACK:
            self._handle_ack(msg, addr)
            
    def _handle_discovery(self, msg: MeshMessage, addr: Tuple[str, int]):
        """Handle discovery beacon from neighbor."""
        if msg.source != self.device_id:
            self.add_neighbor(msg.source, addr[0], addr[1])
            
    def _handle_route_request(self, msg: MeshMessage, addr: Tuple[str, int]):
        """Handle route request (AODV-like)."""
        if msg.destination == self.device_id:
            # We are the destination, send route reply
            reply = MeshMessage(
                msg_type=MessageType.ROUTE_REPLY,
                source=self.device_id,
                destination=msg.source,
                sequence_number=self._get_sequence_number(),
                hop_count=0,
                payload=msg.source.encode("utf-8")
            )
            self._send_to(reply.to_bytes(), msg.source)
        elif msg.hop_count < msg.max_hops:
            # Forward the route request
            msg.hop_count += 1
            msg.previous_hop = self.device_id
            self._broadcast(msg.to_bytes())
            
    def _handle_route_reply(self, msg: MeshMessage, addr: Tuple[str, int]):
        """Handle route reply."""
        if msg.destination == self.device_id:
            # Route reply is for us, extract route
            original_dest = msg.payload.decode("utf-8", errors="ignore")
            route = RouteEntry(
                destination=original_dest,
                next_hop=msg.source,
                hop_count=msg.hop_count + 1,
                sequence_number=msg.sequence_number
            )
            self.routes[original_dest] = route
            logger.info(f"[MESH] Route established to {original_dest} via {msg.source} (hops: {route.hop_count})")
        elif msg.hop_count < msg.max_hops:
            # Forward the route reply
            msg.hop_count += 1
            self._send_to(msg.to_bytes(), msg.destination)
            
    def _handle_data(self, msg: MeshMessage, addr: Tuple[str, int]):
        """Handle data message."""
        if msg.destination == self.device_id:
            # Data is for us, deliver to application
            logger.info(f"[MESH] Received data from {msg.source} (hops: {msg.hop_count})")
            # In real implementation, this would be passed to application callback
        elif msg.hop_count < msg.max_hops:
            # Forward data
            msg.hop_count += 1
            msg.previous_hop = self.device_id
            self._forward_data(msg)
            
    def _handle_ack(self, msg: MeshMessage, addr: Tuple[str, int]):
        """Handle acknowledgment."""
        logger.debug(f"[MESH] Received ACK from {msg.source}")
        
    def _forward_data(self, msg: MeshMessage):
        """Forward data message along route."""
        # Check if we have a route to destination
        if msg.destination in self.routes:
            route = self.routes[msg.destination]
            self._send_to(msg.to_bytes(), route.next_hop)
            route.last_used = time.time()
        else:
            # No route, broadcast to neighbors (flood)
            self._broadcast(msg.to_bytes())
            
    def _send_to(self, data: bytes, device_id: str):
        """Send message to specific device."""
        if device_id in self.neighbors:
            neighbor = self.neighbors[device_id]
            try:
                self.socket.sendto(data, (neighbor.ip, neighbor.port))
            except Exception as e:
                logger.warning(f"[MESH] Failed to send to {device_id}: {e}")
        else:
            logger.warning(f"[MESH] Unknown device: {device_id}")
            
    def _get_sequence_number(self) -> int:
        """Get next sequence number."""
        self.sequence_number += 1
        return self.sequence_number
        
    def _cleanup_stale_neighbors(self):
        """Remove neighbors not seen in 60 seconds."""
        now = time.time()
        stale = [k for k, v in self.neighbors.items() if now - v.last_seen > 60]
        for k in stale:
            del self.neighbors[k]
            logger.debug(f"[MESH] Removed stale neighbor: {k}")
            
    def _cleanup_expired_routes(self):
        """Remove expired routes."""
        now = time.time()
        expired = [k for k, v in self.routes.items() if now > v.expires]
        for k in expired:
            del self.routes[k]
            logger.debug(f"[MESH] Removed expired route to: {k}")
            
    def send_data(self, destination: str, payload: bytes) -> bool:
        """
        Send data to destination through mesh network.
        
        Returns True if data was sent (queued), False if route not found.
        """
        if destination in self.neighbors:
            # Direct neighbor, send immediately
            msg = MeshMessage(
                msg_type=MessageType.DATA,
                source=self.device_id,
                destination=destination,
                sequence_number=self._get_sequence_number(),
                hop_count=0,
                payload=payload
            )
            self._send_to(msg.to_bytes(), destination)
            return True
        elif destination in self.routes:
            # Have a route, send via next hop
            route = self.routes[destination]
            msg = MeshMessage(
                msg_type=MessageType.DATA,
                source=self.device_id,
                destination=destination,
                sequence_number=self._get_sequence_number(),
                hop_count=0,
                payload=payload
            )
            self._send_to(msg.to_bytes(), route.next_hop)
            return True
        else:
            # No route, initiate route discovery
            logger.info(f"[MESH] No route to {destination}, initiating discovery")
            rreq = MeshMessage(
                msg_type=MessageType.ROUTE_REQUEST,
                source=self.device_id,
                destination=destination,
                sequence_number=self._get_sequence_number(),
                hop_count=0,
                max_hops=10
            )
            self._broadcast(rreq.to_bytes())
            self.pending_routes[destination] = time.time()
            return False
            
    def get_topology(self) -> Dict:
        """Get current network topology for visualization."""
        return {
            "device_id": self.device_id,
            "neighbors": [
                {
                    "device_id": n.device_id,
                    "ip": n.ip,
                    "port": n.port,
                    "signal_strength": n.signal_strength,
                    "last_seen": n.last_seen
                }
                for n in self.neighbors.values()
            ],
            "routes": [
                {
                    "destination": r.destination,
                    "next_hop": r.next_hop,
                    "hop_count": r.hop_count,
                    "last_used": r.last_used
                }
                for r in self.routes.values()
            ]
        }

File: esp32/src/mesh.cpp
/**
 * EchoSense Mesh Networking Protocol for ESP32 WROOM
 * AODV-like routing protocol for peer-to-peer device communication
 */

#include "mesh.h"
#include <cstring>

// Constants
#define DISCOVERY_INTERVAL 10000      // 10 seconds
#define CLEANUP_INTERVAL 60000       // 60 seconds
#define NEIGHBOR_TIMEOUT 60000       // 60 seconds
#define ROUTE_TIMEOUT 300000         // 5 minutes
#define MAX_PAYLOAD_SIZE 64
#define UDP_BUFFER_SIZE 128

// MeshMessage serialization
std::vector<uint8_t> MeshMessage::toBytes() const {
    std::vector<uint8_t> data;
    
    // Header: type(1) + source(4) + destination(4) + seq(4) + hop(1) + max_hop(1) + payload_len(1) = 16 bytes
    data.push_back(static_cast<uint8_t>(msgType));
    data.insert(data.end(), source, source + 4);
    data.insert(data.end(), destination, destination + 4);
    
    // Add sequence number (little endian)
    data.push_back(sequenceNumber & 0xFF);
    data.push_back((sequenceNumber >> 8) & 0xFF);
    data.push_back((sequenceNumber >> 16) & 0xFF);
    data.push_back((sequenceNumber >> 24) & 0xFF);
    
    data.push_back(hopCount);
    data.push_back(maxHops);
    data.push_back(payloadLength);
    
    // Add payload
    data.insert(data.end(), payload, payload + payloadLength);
    
    return data;
}

// MeshMessage deserialization
MeshMessage MeshMessage::fromBytes(const uint8_t* data, size_t length) {
    MeshMessage msg;
    
    if (length < 16) return msg;  // Invalid message
    
    size_t offset = 0;
    
    msg.msgType = static_cast<MessageType>(data[offset++]);
    memcpy(msg.source, data + offset, 4);
    offset += 4;
    memcpy(msg.destination, data + offset, 4);
    offset += 4;
    
    msg.sequenceNumber = data[offset] | (data[offset+1] << 8) | (data[offset+2] << 16) | (data[offset+3] << 24);
    offset += 4;
    
    msg.hopCount = data[offset++];
    msg.maxHops = data[offset++];
    msg.payloadLength = data[offset++];
    
    if (length >= offset + msg.payloadLength && msg.payloadLength <= MAX_PAYLOAD_SIZE) {
        memcpy(msg.payload, data + offset, msg.payloadLength);
    }
    
    return msg;
}

// MeshNetwork constructor
MeshNetwork::MeshNetwork(const String& deviceId, uint16_t listenPort)
    : _deviceId(deviceId), _listenPort(listenPort), _running(false),
      _sequenceNumber(0), _lastDiscovery(0), _lastCleanup(0) {
}

// MeshNetwork destructor
MeshNetwork::~MeshNetwork() {
    stop();
}

// Start the mesh network
bool MeshNetwork::begin() {
    if (_running) return true;
    
    if (_udp.begin(_listenPort)) {
        _running = true;
        _lastDiscovery = millis();
        _lastCleanup = millis();
        Serial.printf("[MESH] Started on port %d\n", _listenPort);
        return true;
    }
    
    Serial.println("[MESH] Failed to start UDP");
    return false;
}

// Stop the mesh network
void MeshNetwork::stop() {
    _running = false;
    _udp.stop();
    Serial.println("[MESH] Stopped");
}

// Add a neighbor manually
void MeshNetwork::addNeighbor(const String& deviceId, const String& ip, uint16_t port, float signalStrength) {
    Neighbor neighbor;
    neighbor.deviceId = deviceId;
    neighbor.ip = ip;
    neighbor.port = port;
    neighbor.lastSeen = millis();
    neighbor.signalStrength = signalStrength;
    neighbor.hopCount = 1;
    
    _neighbors[deviceId] = neighbor;
    Serial.printf("[MESH] Added neighbor: %s at %s:%d\n", deviceId.c_str(), ip.c_str(), port);
}

// Send data to destination
bool MeshNetwork::sendData(const String& destination, const uint8_t* data, size_t length) {
    if (length > MAX_PAYLOAD_SIZE) {
        Serial.println("[MESH] Payload too large");
        return false;
    }
    
    // Check if destination is a direct neighbor
    if (_neighbors.count(destination)) {
        MeshMessage msg;
        msg.msgType = DATA;
        strncpy(msg.source, _deviceId.c_str(), 4);
        strncpy(msg.destination, destination.c_str(), 4);
        msg.sequenceNumber = _getSequenceNumber();
        msg.hopCount = 0;
        msg.payloadLength = length;
        memcpy(msg.payload, data, length);
        
        _sendTo(msg.toBytes().data(), msg.toBytes().size(), destination);
        return true;
    }
    
    // Check if we have a route
    if (_routes.count(destination)) {
        RouteEntry& route = _routes[destination];
        MeshMessage msg;
        msg.msgType = DATA;
        strncpy(msg.source, _deviceId.c_str(), 4);
        strncpy(msg.destination, destination.c_str(), 4);
        msg.sequenceNumber = _getSequenceNumber();
        msg.hopCount = 0;
        msg.payloadLength = length;
        memcpy(msg.payload, data, length);
        
        _sendTo(msg.toBytes().data(), msg.toBytes().size(), route.nextHop);
        route.lastUsed = millis();
        return true;
    }
    
    // No route, initiate route discovery
    Serial.printf("[MESH] No route to %s, initiating discovery\n", destination.c_str());
    MeshMessage rreq;
    rreq.msgType = ROUTE_REQUEST;
    strncpy(rreq.source, _deviceId.c_str(), 4);
    strncpy(rreq.destination, destination.c_str(), 4);
    rreq.sequenceNumber = _getSequenceNumber();
    rreq.hopCount = 0;
    rreq.maxHops = 10;
    rreq.payloadLength = 0;
    
    _broadcast(rreq.toBytes().data(), rreq.toBytes().size());
    _pendingRoutes[destination] = millis();
    return false;
}

// Set callback for received data
void MeshNetwork::onDataReceived(DataCallback callback) {
    _dataCallback = callback;
}

// Get current topology
void MeshNetwork::getTopology(std::vector<Neighbor>& neighbors, std::vector<RouteEntry>& routes) {
    neighbors.clear();
    routes.clear();
    
    for (const auto& pair : _neighbors) {
        neighbors.push_back(pair.second);
    }
    
    for (const auto& pair : _routes) {
        routes.push_back(pair.second);
    }
}

// Update loop (call this in main loop)
void MeshNetwork::update() {
    if (!_running) return;
    
    unsigned long now = millis();
    
    // Handle incoming UDP packets
    int packetSize = _udp.parsePacket();
    if (packetSize > 0) {
        uint8_t buffer[UDP_BUFFER_SIZE];
        int len = _udp.read(buffer, sizeof(buffer));
        if (len > 0) {
            String remoteIp = _udp.remoteIP().toString();
            uint16_t remotePort = _udp.remotePort();
            _handleMessage(buffer, len, remoteIp, remotePort);
        }
    }
    
    // Send discovery beacons periodically
    if (now - _lastDiscovery > DISCOVERY_INTERVAL) {
        _sendDiscovery();
        _lastDiscovery = now;
    }
    
    // Cleanup stale neighbors and expired routes
    if (now - _lastCleanup > CLEANUP_INTERVAL) {
        _cleanupStaleNeighbors();
        _cleanupExpiredRoutes();
        _lastCleanup = now;
    }
}

// Get statistics
uint16_t MeshNetwork::getNeighborCount() const {
    return _neighbors.size();
}

uint16_t MeshNetwork::getRouteCount() const {
    return _routes.size();
}

// Send discovery beacon
void MeshNetwork::_sendDiscovery() {
    MeshMessage msg;
    msg.msgType = DISCOVERY;
    strncpy(msg.source, _deviceId.c_str(), 4);
    strncpy(msg.destination, "BCST", 4);
    msg.sequenceNumber = _getSequenceNumber();
    msg.hopCount = 0;
    msg.payloadLength = 0;
    
    std::vector<uint8_t> data = msg.toBytes();
    _broadcast(data.data(), data.size());
    Serial.println("[MESH] Sent discovery beacon");
}

// Broadcast to all neighbors
void MeshNetwork::_broadcast(const uint8_t* data, size_t length) {
    for (const auto& pair : _neighbors) {
        const Neighbor& neighbor = pair.second;
        _udp.beginPacket(neighbor.ip.c_str(), neighbor.port);
        _udp.write(data, length);
        _udp.endPacket();
    }
}

// Send to specific device
void MeshNetwork::_sendTo(const uint8_t* data, size_t length, const String& deviceId) {
    if (_neighbors.count(deviceId)) {
        const Neighbor& neighbor = _neighbors[deviceId];
        _udp.beginPacket(neighbor.ip.c_str(), neighbor.port);
        _udp.write(data, length);
        _udp.endPacket();
    } else {
        Serial.printf("[MESH] Unknown device: %s\n", deviceId.c_str());
    }
}

// Handle incoming message
void MeshNetwork::_handleMessage(const uint8_t* data, size_t length, const String& remoteIp, uint16_t remotePort) {
    MeshMessage msg = MeshMessage::fromBytes(data, length);
    String source = _extractDeviceId(msg.source);
    
    // Update neighbor if this is from a direct neighbor
    if (source != _deviceId && msg.hopCount == 0) {
        addNeighbor(source, remoteIp, remotePort);
    }
    
    // Handle based on message type
    switch (msg.msgType) {
        case DISCOVERY:
            _handleDiscovery(msg, remoteIp, remotePort);
            break;
        case ROUTE_REQUEST:
            _handleRouteRequest(msg);
            break;
        case ROUTE_REPLY:
            _handleRouteReply(msg);
            break;
        case DATA:
            _handleData(msg);
            break;
        case ACK:
            // Handle acknowledgment
            Serial.printf("[MESH] Received ACK from %s\n", source.c_str());
            break;
        default:
            Serial.println("[MESH] Unknown message type");
            break;
    }
}

// Handle discovery message
void MeshNetwork::_handleDiscovery(const MeshMessage& msg, const String& remoteIp, uint16_t remotePort) {
    String source = _extractDeviceId(msg.source);
    if (source != _deviceId) {
        addNeighbor(source, remoteIp, remotePort);
    }
}

// Handle route request
void MeshNetwork::_handleRouteRequest(const MeshMessage& msg) {
    String destination = _extractDeviceId(msg.destination);
    
    if (destination == _deviceId) {
        // We are the destination, send route reply
        MeshMessage reply;
        reply.msgType = ROUTE_REPLY;
        strncpy(reply.source, _deviceId.c_str(), 4);
        strncpy(reply.destination, _extractDeviceId(msg.source).c_str(), 4);
        reply.sequenceNumber = _getSequenceNumber();
        reply.hopCount = 0;
        reply.payloadLength = msg.source[0];  // Store original destination in first byte
        memcpy(reply.payload, msg.source, 4);
        
        std::vector<uint8_t> data = reply.toBytes();
        _sendTo(data.data(), data.size(), _extractDeviceId(msg.source));
    } else if (msg.hopCount < msg.maxHops) {
        // Forward the route request
        MeshMessage forwardMsg = msg;
        forwardMsg.hopCount++;
        std::vector<uint8_t> data = forwardMsg.toBytes();
        _broadcast(data.data(), data.size());
    }
}

// Handle route reply
void MeshNetwork::_handleRouteReply(const MeshMessage& msg) {
    String destination = _extractDeviceId(msg.destination);
    
    if (destination == _deviceId) {
        // Route reply is for us
        String originalDest = _extractDeviceId(reinterpret_cast<const char*>(msg.payload));
        
        RouteEntry route;
        route.destination = originalDest;
        route.nextHop = _extractDeviceId(msg.source);
        route.hopCount = msg.hopCount + 1;
        route.sequenceNumber = msg.sequenceNumber;
        route.lastUsed = millis();
        route.expires = millis() + ROUTE_TIMEOUT;
        
        _routes[originalDest] = route;
        Serial.printf("[MESH] Route established to %s via %s (hops: %d)\n", 
                      originalDest.c_str(), route.nextHop.c_str(), route.hopCount);
    } else if (msg.hopCount < msg.maxHops) {
        // Forward the route reply
        MeshMessage forwardMsg = msg;
        forwardMsg.hopCount++;
        std::vector<uint8_t> data = forwardMsg.toBytes();
        _sendTo(data.data(), data.size(), destination);
    }
}

// Handle data message
void MeshNetwork::_handleData(const MeshMessage& msg) {
    String destination = _extractDeviceId(msg.destination);
    
    if (destination == _deviceId) {
        // Data is for us, deliver to application
        String source = _extractDeviceId(msg.source);
        Serial.printf("[MESH] Received data from %s (hops: %d)\n", source.c_str(), msg.hopCount);
        
        if (_dataCallback) {
            _dataCallback(source, msg.payload, msg.payloadLength, msg.hopCount);
        }
    } else if (msg.hopCount < msg.maxHops) {
        // Forward data
        _forwardData(msg);
    }
}

// Forward data message
void MeshNetwork::_forwardData(const MeshMessage& msg) {
    String destination = _extractDeviceId(msg.destination);
    
    // Check if we have a route to destination
    if (_routes.count(destination)) {
        RouteEntry& route = _routes[destination];
        MeshMessage forwardMsg = msg;
        forwardMsg.hopCount++;
        
        std::vector<uint8_t> data = forwardMsg.toBytes();
        _sendTo(data.data(), data.size(), route.nextHop);
        route.lastUsed = millis();
    } else {
        // No route, broadcast to neighbors (flood)
        MeshMessage forwardMsg = msg;
        forwardMsg.hopCount++;
        std::vector<uint8_t> data = forwardMsg.toBytes();
        _broadcast(data.data(), data.size());
    }
}

// Cleanup stale neighbors
void MeshNetwork::_cleanupStaleNeighbors() {
    unsigned long now = millis();
    std::vector<String> stale;
    
    for (const auto& pair : _neighbors) {
        if (now - pair.second.lastSeen > NEIGHBOR_TIMEOUT) {
            stale.push_back(pair.first);
        }
    }
    
    for (const String& deviceId : stale) {
        _neighbors.erase(deviceId);
        Serial.printf("[MESH] Removed stale neighbor: %s\n", deviceId.c_str());
    }
}

// Cleanup expired routes
void MeshNetwork::_cleanupExpiredRoutes() {
    unsigned long now = millis();
    std::vector<String> expired;
    
    for (const auto& pair : _routes) {
        if (now > pair.second.expires) {
            expired.push_back(pair.first);
        }
    }
    
    for (const String& destination : expired) {
        _routes.erase(destination);
        Serial.printf("[MESH] Removed expired route to: %s\n", destination.c_str());
    }
}

// Get next sequence number
uint32_t MeshNetwork::_getSequenceNumber() {
    return ++_sequenceNumber;
}

// Extract device ID from 4-byte array
String MeshNetwork::_extractDeviceId(const char* deviceId) {
    char id[5];
    memcpy(id, deviceId, 4);
    id[4] = '\0';
    return String(id);
}
